drivers/usb/usbuhci_new/roothub.c

```diff
@@ -0,0 +1,470 @@
+    }
+
+    if (UhciExtension->ResetChangePortMask & port)
+    {
+        portChange.ResetChange = TRUE;
+    }
+
+    PortStatus->PortStatus.Usb20PortStatus = portStatus;
+    PortStatus->PortChange.Usb20PortChange = portChange;
+
+    //DPRINT("UhciRHGetPortStatus: PortControl.AsUSHORT[%x] - %X, PortStatus - %X\n",
+    //       Port,
+    //       PortControl.AsUSHORT,
+    //       PortStatus->AsUlong32);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHGetHubStatus(IN PVOID uhciExtension,
+                   IN PUSB_HUB_STATUS_AND_CHANGE HubStatus)
+{
+    //DPRINT("UhciRHGetHubStatus: ...\n");
+    HubStatus->AsUlong32 = 0;
+    return MP_STATUS_SUCCESS;
+}
+
+VOID
+NTAPI
+UhciRHPortResetComplete(IN PVOID uhciExtension,
+                        IN PVOID pPort)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    ULONG ix;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+    USHORT Port;
+
+    DPRINT("UhciRHPortResetComplete: ...\n");
+
+    BaseRegister = UhciExtension->BaseRegister;
+    Port = *(PUSHORT)pPort - 1;
+
+    PortControlRegister = &BaseRegister->PortControl[Port].AsUSHORT;
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    PortControl.ConnectStatusChange = FALSE;
+    PortControl.PortEnableDisableChange = FALSE;
+    PortControl.PortReset = FALSE;
+
+    WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+
+    while (UhciHardwarePresent(UhciExtension))
+    {
+        PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+        if (PortControl.PortReset == TRUE)
+        {
+            continue;
+        }
+
+        for (ix = 0; ix < 10; ++ix)
+        {
+            KeStallExecutionProcessor(50);
+
+            PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+            if (PortControl.PortEnabledDisabled == TRUE)
+            {
+                break;
+            }
+
+            PortControl.PortEnabledDisabled = TRUE;
+            WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+        }
+
+        PortControl.ConnectStatusChange = TRUE;
+        PortControl.PortEnableDisableChange = TRUE;
+        WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+
+        if (UhciExtension->HcFlavor == UHCI_VIA &&
+            UhciExtension->HcFlavor == UHCI_VIA_x01 &&
+            UhciExtension->HcFlavor == UHCI_VIA_x02 &&
+            UhciExtension->HcFlavor == UHCI_VIA_x03 &&
+            UhciExtension->HcFlavor == UHCI_VIA_x04)
+        {
+            DPRINT("UhciRHPortResetComplete: Via chip. FIXME\n");
+            DbgBreakPoint();
+            return;
+        }
+
+        UhciExtension->ResetChangePortMask |= (1 << Port);
+        UhciExtension->ResetPortMask &= ~(1 << Port);
+
+        RegPacket.UsbPortInvalidateRootHub(UhciExtension);
+
+        return;
+    }
+}
+
+VOID
+NTAPI
+UhciRHSetFeaturePortResetWorker(IN PUHCI_EXTENSION UhciExtension,
```

Comment:
I'm curious why this is its own function. It only seems to be called in once place.

Reply:
It use also UhciVia...(), but it not implemented.